src/components/common/PaginatedTxns/index.tsx

```diff
@@ -9,93 +11,78 @@ import InfiniteScroll from '../InfiniteScroll'
 import SkeletonTxList from './SkeletonTxList'
 import BatchExecuteButton from '@/components/transactions/BatchExecuteButton'
 import TxFilterButton from '@/components/transactions/TxFilterButton'
-import { TxFilter, useTxFilter } from '@/utils/tx-history-filter'
-import { isTransactionListItem } from '@/utils/transaction-guards'
-import type { TransactionListPage } from '@gnosis.pm/safe-react-gateway-sdk'
+import { type TxFilter, useTxFilter } from '@/utils/tx-history-filter'
 import { BatchExecuteHoverProvider } from '@/components/transactions/BatchExecuteButton/BatchExecuteHoverProvider'
+import { isTransactionListItem } from '@/utils/transaction-guards'
 
 const NoQueuedTxns = () => (
   <Box mt="5vh">
     <PagePlaceholder imageUrl="/images/no-transactions.svg" text="Queued transactions will appear here" />
   </Box>
 )
 
-const getResultCount = (filter: TxFilter, page: TransactionListPage) => {
+const getFilterResultCount = (filter: TxFilter, page: TransactionListPage) => {
   const count = page.results.filter(isTransactionListItem).length
 
   return `${page.next ? '> ' : ''}${count} ${filter.type} transactions found`.toLowerCase()
 }
 
-const TxPage = ({
-  pageUrl,
-  useTxns,
-  onNextPage,
-  isFirstPage,
-}: {
-  pageUrl: string
-  useTxns: typeof useTxHistory | typeof useTxQueue
-  onNextPage?: (pageUrl?: string) => void
-  isFirstPage: boolean
-}): ReactElement => {
-  const { page, error } = useTxns(pageUrl)
+const PaginatedTxns = ({ useTxns }: { useTxns: typeof useTxHistory | typeof useTxQueue }): ReactElement => {
+  const [pages, setPages] = useState<Array<{ pageUrl?: string; data: TransactionListPage }>>([])
+  const [pageUrl, setPageUrl] = useState<string>()
   const [filter] = useTxFilter()
-
   const isQueue = useTxns === useTxQueue
+  const allItems = useMemo(() => flatten(pages.map((page) => page.data.results)), [pages])
 
-  if (page?.results) {
-    return (
-      <>
-        {/* FIXME: batching will only work for the first page results */}
-        {isFirstPage && (
-          <Box display="flex" flexDirection="column" alignItems="flex-end" mt={['-94px', '-44px']} mb={['60px', 0]}>
-            {isQueue ? <BatchExecuteButton items={page.results} /> : <TxFilterButton />}
-            {filter && <Typography mt={2}>{getResultCount(filter, page)}</Typography>}
-          </Box>
-        )}
-
-        {page.results.length ? <TxList items={page.results} /> : isQueue && <NoQueuedTxns />}
-
-        {onNextPage && page.next && (
-          <Box my={4} textAlign="center">
-            <InfiniteScroll onLoadMore={() => onNextPage(page.next)} />
-          </Box>
-        )}
-      </>
-    )
-  }
-
-  if (error) {
-    return <ErrorMessage>Error loading transactions</ErrorMessage>
-  }
-
-  return <SkeletonTxList />
-}
+  // Reset the pages when the filter changes
+  useEffect(() => {
+    setPages([])
+    setPageUrl(undefined)
+  }, [filter, useTxns])
```

Comment:
I've noticed this doesn't always reset the pageUrl quick enough. Sometimes a filtered page loads quickly before pageUrl is reset, and it's not the first page.

Reply:
Is there a reason we are relying on `useTxns` as a dependency here? It seems to update correctly when I remove it.

Reply:
I wasn't sure If it re-mounts the entire component when you switch tabs between Queue and History.

Reply:
Makes total sense, `useTxns` also doesn't seem to change when applying the filter so there must be another reason.